mozetl/maudau/maudau.py

```diff
@@ -0,0 +1,214 @@
+    max_submission_date = end_date + ACTIVITY_SUBMISSION_LAG
+    max_submission_string = format_as_submission_date(
+        max_submission_date.isoformat())
+    return min_activity_iso, max_activity_iso, min_submission_string, \
+        max_submission_string
+
+
+def count_active_users(frame, end_date, days_back):
+    """
+    Tally active users according to externally-defined heuristics,
+    specifically client-reported subsession_start_date.
+
+    :frame DataFrame(): conforming to main_summary schema
+    :end_date DT.date(): the last day on which to count activity.
+    :days_back int: how far back to go; 0 for daily and 28 for monthly.
+
+    Note that return values are not stable over time.
+
+    See https://bugzilla.mozilla.org/show_bug.cgi?id=1240849
+    """
+    params = generate_filter_parameters(end_date, days_back)
+    (min_activity_iso, max_activity_iso,
+     min_submission_string, max_submission_string) = params
+
+    filtered = filter_date_range(
+        frame,
+        frame.subsession_start_date,
+        min_activity_iso,
+        max_activity_iso,
+        frame.submission_date_s3,
+        min_submission_string,
+        max_submission_string)
+    return count_distinct_clientids(filtered)
+
+
+def parse_last_rollup(basename, start_date=None):
+    """
+    Extract the valid old counts from basename and return them with the
+    first date that needs to be re-counted.
+    """
+    start_date = start_date or today
+    since = (start_date - DT.timedelta(INSTABLE_WINDOW))
+    since = format_as_submission_date(since.isoformat())
+    carryover = []
+    with open(basename) as f:
+        reader = csv.DictReader(f)
+        for row in reader:
+            if row['day'] >= since:
+                break
+            carryover.append(row)
+    return since, carryover
+
+
+def get_last_rollup(transferer):
+    '''
+    Always chose the latest rollup, even if start_date is old.
+    '''
+    basename = MAUDAU_ROLLUP_BASENAME
+    key = os.path.join(STORAGE_SUB_DIR, basename)
+    try:
+        transferer.download_file(
+            STORAGE_BUCKET,
+            key,
+            basename)
+    except botocore.exceptions.ClientError as e:
+        # If the file wasn't there, that's ok. Otherwise, abort!
+        if e.response['Error']['Code'] != "404":
+            raise e
+        else:
+            msg = "Did not find an existing file at '{}'".format(basename)
+            logger.info(msg)
+        return None
+    else:
+        return basename
+
+
+def generate_counts(frame, since, until=None):
+    cols = [frame.client_id.alias('clientId'),
+            'subsession_start_date', 'submission_date_s3']
+    narrow = frame.select(cols)
+    updates = []
+    generated = format_as_submission_date(today.isoformat())
+    start = DT.datetime.strptime(since, '%Y%m%d').date()
+    until = (DT.datetime.strptime(until, '%Y%m%d').date() if until
+             else today)
+    while start < until:
```

Comment:
Is this functionally equivalent to a sliding window over a day and a month, each with a stride of 1?